### PR TITLE
[BEAM-2822] Add fn API progress reporting protos

### PR DESCRIPTION
**Pull request by vikkyrk**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by vikkyrk** [comment]
R: @lukecwik @robertwb : Tried to bring it inline with all the comments in the doc. PTAL.

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto [hunk 1]
```diff
@@ -197,15 +198,86 @@ message ProcessBundleProgressRequest {
...
+      Measured measured = 1;
+
+      // Estimated metrics.
+
+      // (Optional) Sum of estimated fraction of known work remaining for all
+      // active elements, as reported by this transform.
+      // If not reported, a Runner could extrapolate this from the processed
+      // elements.
+      // TODO: Handle the case when known work is infinite.
+      double fractions_remaining = 2;
```
Comment (lukecwik) [inline sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: fractions_remaining seems to imply that this is a repeated field, should we name this fraction_remaining?

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto [hunk 2]
```diff
@@ -197,15 +198,86 @@ message ProcessBundleProgressRequest {
...
 message ProcessBundleProgressResponse {
-  // (Required) The finished amount of work. A monotonically increasing
-  // unitless measure of work finished.
-  double finished_work = 1;
-
-  // (Required) The known amount of backlog for the process bundle request.
-  // Computed as:
-  //   (estimated known work - finish work) / finished work
-  double backlog = 2;
+  Metrics metrics = 1;
```
Comment (lukecwik) [inline sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: add a comment saying that this field is required.

Also, add a metrics field to the ProcessBundleResponse.

**Comment by lukecwik** [comment]
LGTM if you agree with my comments.

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto (hunk 2 above)
Comment (vikkyrk) [reply sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: Marked it as required. I had added metrics field to `ProcessBundleResponse`.

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto (hunk 1 above)
Comment (vikkyrk) [reply sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: Done

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto [hunk 3]
```diff
@@ -189,6 +189,7 @@ message ProcessBundleRequest {
 
 // Stable
 message ProcessBundleResponse {
+  // (Required)
```
Comment (lukecwik) [inline sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: I wouldn't say that this is required, I think a comment stating that if metrics is supported by the SDK harness, then this response here represents the 'final' metrics to record for this bundle.

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto [hunk 4]
```diff
@@ -277,6 +278,7 @@ message Metrics {
 }
 
 message ProcessBundleProgressResponse {
+  // (Required)
```
Comment (lukecwik) [inline sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: I would mention within the InstructionRequest, that for any unsupported request type, an error should be returned with a matching id and then mention on the ProcessBundleProgressRequest that this is an optional request to be handled and is used to support advanced SDK features such as SplittableDoFn.

Comment (vikkyrk) [reply sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: Done

File: sdks/common/fn-api/src/main/proto/beam_fn_api.proto (hunk 3 above)
Comment (vikkyrk) [reply sdks/common/fn-api/src/main/proto/beam_fn_api.proto]: I was going for empty metrics response but this makes more sense.

**Comment by vikkyrk** [comment]
Ping for merge.

**Comment by lukecwik** [comment]
LGTM, merging

**Pull request closed without merging**